### PR TITLE
Handle animations of models

### PR DESCRIPTION
**Pull request by AlaricBaraou**
Change introduced in this MR

- Automatically run the first animation if present when model load
- Display a folder specific to animations in Leva if animations are present
- Run / stop each animation through Leva controls

I did some quick test and it looks like it doesn't affect models without animations.

Feel free to suggest change / optimise the code.
I honestly felt rusty with react and I might have missed some good practice with hooks or others 😅

### CODE REVIEW

File: src/components/canvas/Model.js [hunk 1]
```diff
@@ -12,21 +13,51 @@ const Model = ({ file }) => {
...
   useEffect(() => {
     scene.traverse((obj) => {
       if (obj.isMesh && obj.material) {
         obj.material.wireframe = controls.wireframe
       }
     })
+
+    if(animations[0]){
+      defaultAnimationsControls[animations[0].name] = true
+    }
```
Comment (njm222) [inline src/components/canvas/Model.js:54]: seems like this will break if animations is empty

File: src/components/canvas/Model.js [hunk 2]
```diff
@@ -12,21 +13,51 @@ const Model = ({ file }) => {
     ...lightControls,
   })
 
-  const { scene } = useGLTF(file)
+  const { animations,scene } = useGLTF(file)
+  const content = scene || scenes[0];
```
Comment (njm222) [inline src/components/canvas/Model.js:17]: scenes is undefined

File: src/components/canvas/Model.js [hunk 3]
```diff
@@ -12,21 +13,51 @@ const Model = ({ file }) => {
...
+    defaultAnimationsControls
+  ))
+
+  for(let clipName in animationsClip){
+    if(controlsAnim[clipName]){
+      animationsClip[clipName].play();
+    }else{
+      animationsClip[clipName].stop();
+    }
+  }
```
Comment (njm222) [inline src/components/canvas/Model.js:37]: all of this can be wrapped in a useMemo and should be cleaned up.. hard to follow what is going on

**Review by njm222 — CHANGES_REQUESTED** [review]
don't forget to lint before commiting next time!

File: src/components/canvas/Model.js (hunk 1 above)
Comment (AlaricBaraou) [reply src/components/canvas/Model.js]: It didn't seem to break before but to be sure I changed it.
animations from useGLTF is always an Array even when there is not animations property in the .gltf

File: src/components/canvas/Model.js (hunk 2 above)
Comment (AlaricBaraou) [reply src/components/canvas/Model.js]: fixed 🙇‍♂️

File: src/components/canvas/Model.js (hunk 3 above)
Comment (AlaricBaraou) [reply src/components/canvas/Model.js]: I wrapped as much as I could in the latest commit but as far as I know I can't put the useControls hook inside the useMemo

For readability, I could add comments next, not sure how to improve it more.

Let me know what you think.

File: src/components/canvas/Model.js (hunk 1 above)
Comment (SaraVieira) [reply src/components/canvas/Model.js]: Yap, but you can check animations.length instead if you wanna make it fancier

Comment (AlaricBaraou) [reply src/components/canvas/Model.js]: Yes that's what I did in the latest commit :)

File: src/components/canvas/Model.js [hunk 4]
```diff
@@ -13,26 +14,32 @@ const Model = ({ file }) => {
...
+    'Animations',
+    () => defaultAnimationsControls
+  )
+
+  for (let clipName in animationsClip) {
+    if (controlsAnim[clipName]) {
+      animationsClip[clipName].play()
+    } else {
+      animationsClip[clipName].stop()
     }
```
Comment (njm222) [inline src/components/canvas/Model.js]: This should be wrapped in a useEffect so that it doesn't get called unnecessarily

**Review by njm222 — CHANGES_REQUESTED** [review]

File: src/components/canvas/Model.js (hunk 4 above)
Comment (AlaricBaraou) [reply src/components/canvas/Model.js]: Thanks you're right, really makes sense once you say it.

**Review by njm222 — CHANGES_REQUESTED** [review]
Looks much better now thanks!

I hope I'm not being too picky but for readability:

1. Can you change those for loops to .forEach since we don't need to interrupt the loop ( no break / continue )
2. variable name changes:
`animationsClip` => `animationClips`
`controlsAnim` => `animationsControl`
3. can you move this import to the top?`import { useMemo } from 'react'`

**Comment by AlaricBaraou** [comment]
> I hope I'm not being too picky but for readability:

No that's great, I'm learning from feedbacks

**Comment by SaraVieira** [comment]
Thank you so much!! This looks amazing! No more comments! Just a thank you 🎉
